File: frontend/src/api/sync.ts
import api from "@/utils/api";

// 同步状态枚举
export enum SyncStatus {
  IDLE = "idle",
  SYNCING = "syncing",
  SUCCESS = "success",
  FAILED = "failed",
  CONFLICT = "conflict",
}

// 冲突解决策略
export enum ConflictResolution {
  MANUAL = "manual",
  AUTO_LOCAL = "auto_local",
  AUTO_REMOTE = "auto_remote",
  SMART_MERGE = "smart_merge",
}

// 数据类型定义
export interface GitHubSyncConfig {
  repository: string;
  branch: string;
  auto_sync: boolean;
  sync_interval: number;
  include_files: string[];
  exclude_files: string[];
  conflict_resolution: ConflictResolution;
  last_sync?: string;
  status: SyncStatus;
}

export interface GitHubSyncConfigRequest {
  repository: string;
  token: string;
  branch?: string;
  auto_sync?: boolean;
  sync_interval?: number;
  include_files?: string[];
  exclude_files?: string[];
  conflict_resolution?: ConflictResolution;
}

export interface FileChangeInfo {
  file_path: string;
  file_type: "added" | "modified" | "deleted";
  size?: number;
  hash?: string;
  last_modified?: string;
}

export interface SyncStatusResponse {
  status: SyncStatus;
  last_sync?: string;
  next_sync?: string;
  pending_files: FileChangeInfo[];
  current_operation?: any;
  progress?: number;
  message?: string;
  total_files: number;
  synced_files: number;
}

export interface SyncHistoryItem {
  timestamp: string;
  operation_type: "manual_sync" | "auto_sync" | "restore" | "conflict_resolve";
  status: SyncStatus;
  files_count: number;
  message?: string;
  duration?: number;
}

export interface TestConnectionResponse {
  success: boolean;
  message: string;
  repository_info?: {
    name: string;
    full_name: string;
    private: boolean;
    default_branch: string;
  };
}

export interface ManualSyncRequest {
  force?: boolean;
  files?: string[];
}

export interface RestoreRequest {
  timestamp?: string;
  commit_hash?: string;
  force?: boolean;
}

// API 接口
export const syncAPI = {
  // 配置同步
  configureSync: (config: GitHubSyncConfigRequest): Promise<GitHubSyncConfig> =>
    api.post("/sync/config", config),

  // 获取同步配置
  getSyncConfig: (): Promise<GitHubSyncConfig> => api.get("/sync/config"),

  // 测试GitHub连接
  testConnection: (): Promise<TestConnectionResponse> =>
    api.post("/sync/test-connection"),

  // 获取同步状态
  getSyncStatus: (): Promise<SyncStatusResponse> => api.get("/sync/status"),

  // 手动同步
  manualSync: (
    params?: ManualSyncRequest
  ): Promise<{ success: boolean; message: string }> =>
    api.post("/sync/manual", params || {}),

  // 从GitHub恢复数据
  restoreFromGitHub: (
    params?: RestoreRequest
  ): Promise<{ success: boolean; message: string }> =>
    api.post("/sync/restore", params || {}),

  // 获取同步历史
  getSyncHistory: (
    page = 1,
    pageSize = 20
  ): Promise<{
    history: SyncHistoryItem[];
    total_count: number;
    page: number;
    page_size: number;
  }> => api.get("/sync/history", { params: { page, page_size: pageSize } }),

  // 删除同步配置
  removeSyncConfig: (): Promise<{ success: boolean; message: string }> =>
    api.delete("/sync/config"),

  // 暂停自动同步
  pauseAutoSync: (): Promise<{ success: boolean; message: string }> =>
    api.post("/sync/pause"),

  // 恢复自动同步
  resumeAutoSync: (): Promise<{ success: boolean; message: string }> =>
    api.post("/sync/resume"),

  // 获取仓库分支列表
  getRepositoryBranches: (): Promise<{
    branches: string[];
    default_branch: string;
  }> => api.get("/sync/branches"),

  // 获取最近提交
  getRecentCommits: (
    limit = 10
  ): Promise<{
    commits: Array<{
      sha: string;
      message: string;
      author: string;
      date: string;
      url: string;
    }>;
  }> => api.get("/sync/commits", { params: { limit } }),
};

File: backend/app/routers/sync.py
from fastapi import APIRouter, HTTPException, Depends
from typing import Optional, List
from sqlalchemy.orm import Session
from app.database import get_db

from app.models.sync_schemas import (
    GitHubSyncConfigRequest, GitHubSyncConfigResponse, SyncStatusResponse,
    SyncHistoryResponse, SyncHistoryItem, ManualSyncRequest, RestoreRequest, TestConnectionResponse,
    ConflictResolutionRequest, SyncMetrics
)
from app.services.github_sync_service import GitHubSyncService
from app.utils.auth import get_current_user
from app.models.github_sync import GitHubSync

router = APIRouter(tags=["同步管理"])

def get_sync_service(db: Session = Depends(get_db)) -> GitHubSyncService:
    """Dependency to get a GitHubSyncService instance with a db session."""
    return GitHubSyncService(db=db)


@router.post("/config", response_model=GitHubSyncConfigResponse)
async def configure_sync(
    config_request: GitHubSyncConfigRequest,
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """配置GitHub同步设置"""
    try:
        config = await sync_service.configure_sync(config_request.model_dump())
        
        status = await sync_service.get_sync_status()
        
        # We need to build the response model correctly
        response_data = config.model_dump()
        response_data['status'] = status.status
        response_data['last_sync'] = status.last_sync
        
        return GitHubSyncConfigResponse(**response_data)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"配置同步失败: {str(e)}")


@router.get("/config", response_model=GitHubSyncConfigResponse)
async def get_sync_config(
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """获取当前同步配置"""
    config = await sync_service.get_config()
    
    if not config:
        raise HTTPException(status_code=404, detail="同步配置未设置")
    
    status = await sync_service.get_sync_status()
    
    response_data = config.model_dump()
    response_data['status'] = status.status
    response_data['last_sync'] = status.last_sync

    return GitHubSyncConfigResponse(**response_data)


@router.post("/test-connection", response_model=TestConnectionResponse)
async def test_github_connection(current_user: dict = Depends(get_current_user)):
    """测试GitHub连接"""
    try:
        result = await get_sync_service().test_connection()
        return TestConnectionResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"测试连接失败: {str(e)}")


@router.get("/status", response_model=SyncStatusResponse)
async def get_sync_status(
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """获取同步状态"""
    try:
        status = await sync_service.get_sync_status()
        return status
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取同步状态失败: {str(e)}")


@router.post("/manual")
async def manual_sync(
    request: ManualSyncRequest,
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """手动触发同步"""
    try:
        result = await sync_service.manual_sync(
            force=request.force,
            files=request.files
        )
        return {"success": result, "message": "同步完成" if result else "同步失败"}
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"同步失败: {str(e)}")


@router.post("/restore")
async def restore_from_github(
    request: RestoreRequest,
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """从GitHub恢复数据"""
    try:
        result = await sync_service.restore_from_github(
            commit_hash=request.commit_hash,
            force=request.force
        )
        return {"success": result, "message": "恢复完成" if result else "恢复失败"}
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"恢复失败: {str(e)}")


@router.get("/history", response_model=SyncHistoryResponse)
async def get_sync_history(
    page: int = 1,
    page_size: int = 20,
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """获取同步历史"""
    if page < 1 or page_size < 1 or page_size > 100:
        raise HTTPException(status_code=400, detail="无效的分页参数")
    
    result = await sync_service.get_sync_history(page, page_size)
    return result


@router.post("/resolve-conflict")
async def resolve_conflict(
    request: ConflictResolutionRequest,
    current_user: dict = Depends(get_current_user)
):
    """解决同步冲突"""
    try:
        # TODO: 实现冲突解决逻辑
        return {
            "success": True,
            "message": f"冲突解决完成: {request.file_path}"
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"解决冲突失败: {str(e)}")


@router.get("/metrics", response_model=SyncMetrics)
async def get_sync_metrics(current_user: dict = Depends(get_current_user)):
    """获取同步指标"""
    try:
        # TODO: 实现指标统计
        return SyncMetrics(
            total_syncs=0,
            successful_syncs=0,
            failed_syncs=0,
            last_sync_duration=None,
            average_sync_duration=None,
            data_synced_mb=0.0,
            conflicts_resolved=0
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取指标失败: {str(e)}")


@router.delete("/config")
async def remove_sync_config(
    current_user: dict = Depends(get_current_user),
    sync_service: GitHubSyncService = Depends(get_sync_service)
):
    """删除同步配置"""
    try:
        db_config = sync_service.db.query(GitHubSync).first()
        if db_config:
            sync_service.db.delete(db_config)
            sync_service.db.commit()
            # Reset in-memory config
            sync_service._config = None
            sync_service._github_client = None
        
        return {"success": True, "message": "同步配置已删除"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"删除配置失败: {str(e)}")


@router.post("/pause")
async def pause_auto_sync(current_user: dict = Depends(get_current_user)):
    """暂停自动同步"""
    try:
        await get_sync_service()._ensure_initialized()
        await get_sync_service()._stop_file_watcher()
        
        if get_sync_service()._config:
            get_sync_service()._config.auto_sync = False
            await get_sync_service()._save_config()
        
        return {"success": True, "message": "自动同步已暂停"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"暂停自动同步失败: {str(e)}")


@router.post("/resume")
async def resume_auto_sync(current_user: dict = Depends(get_current_user)):
    """恢复自动同步"""
    try:
        await get_sync_service()._ensure_initialized()
        if get_sync_service()._config:
            get_sync_service()._config.auto_sync = True
            await get_sync_service()._save_config()
            await get_sync_service()._start_file_watcher()
        
        return {"success": True, "message": "自动同步已恢复"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"恢复自动同步失败: {str(e)}")


@router.get("/branches")
async def get_repository_branches(current_user: dict = Depends(get_current_user)):
    """获取仓库分支列表"""
    try:
        if not get_sync_service()._github_client or not get_sync_service()._config:
            raise HTTPException(status_code=400, detail="GitHub连接未配置")
        
        repo = get_sync_service()._github_client.get_repo(get_sync_service()._config.repository)
        branches = [branch.name for branch in repo.get_branches()]
        
        return {
            "branches": branches,
            "default_branch": repo.default_branch
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"获取分支列表失败: {str(e)}")


@router.get("/commits")
async def get_recent_commits(
    limit: int = 10,
    current_user: dict = Depends(get_current_user)
):
    """获取最近的提交记录"""
    try:
        if not get_sync_service()._github_client or not get_sync_service()._config:
            raise HTTPException(status_code=400, detail="GitHub连接未配置")
        
        if limit < 1 or limit > 50:
            raise HTTPException(status_code=400, detail="limit参数必须在1-50之间")
        
        repo = get_sync_service()._github_client.get_repo(get_sync_service()._config.repository)
        commits = repo.get_commits(sha=get_sync_service()._config.branch)[:limit]
        
        commit_list = []
        for commit in commits:
            commit_list.append({
                "sha": commit.sha,
                "message": commit.commit.message,
                "author": commit.commit.author.name,
                "date": commit.commit.author.date.isoformat(),
                "url": commit.html_url
            })
        
        return {"commits": commit_list}
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"获取提交记录失败: {str(e)}")
